### ISSUE TITLE
[FEATURE] multiclass with dask run each class on separate GPUs

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#6721]
@trivialfis  just curious how dask works currently.  If have 8 GPUs and doing 8 class problem, will it be smart and train each class on each GPU?  Or does it build all classes on all GPUs?  I guess a feature request would be to do the optimal thing with dask if multiclass case, by not distributing data if each GPU can handle the data, so get pure speed boost from each GPU doing its own class building.

**Comment by RAMitchell** [comment]
If you distributed by class each worker still needs the entire X matrix, you could only distribute the labels. I don't think there is a benefit to optimising for this case.

**Comment by pseudotensor** [comment]
There are primarily two kinds of parallelism and dask currently only benefits from data parallelism.  In general dask version or prior version do not try to take advantage of compute parallelism and so scale quite poorly with number of GPUs.

In balanced case, the labels distribute perfectly across GPUs.  Only in heavily imbalanced case would it not be helpful in terms of data paralellism.

In terms of compute parallelism, regardless of data parallism, the work can be distributed in a nearly embarassingly parallel way for large number of classes, achieving a much better scalability than currently exists.

In terms of using both data and compute parallelism, the benefits are quite clear.  I'm not sure what you mean by "could only distribute by the labels".  That would be the great case, especially for generic balanced case.  When doing each label on each GPU for each class, one obtains both less data on each GPU as well as more quickly finishing since one doesn't have to do the other classes.  So I'm pretty confused by  your response that you think there is no benefit.

**Comment by trivialfis** [comment]
> one obtains both less data on each GPU as well as more quickly finishing

The number of data on each GPU would be the same (the whole dataset).  But yes it can finish more quickly.

**Comment by pseudotensor** [comment]
> > one obtains both less data on each GPU as well as more quickly finishing
> 
> The number of data on each GPU would be the same (the whole dataset). But yes it can finish more quickly.

Can you explain?  I thought each tree is independently trained for a given class, so only the labels for a given class are needed.  E.g., if it's 8 classes, then the only thing one needs is that it is the target class or not.  You don't need to keep all 8 classes on all GPUs.

**Comment by trivialfis** [comment]
Given C classes, the label `y_i` is passed as value `y_i \in [0, C - 1]`, and `X` is passed as a whole matrix.  Then we calculate gradient for each class by using derivative of softmax.  After that, we train 1 tree for gradient from each class.  So the data is one single block, gradient and prediction are separated for each class.

**Comment by trivialfis** [comment]
So it won't make it memory efficient, but with the whole dataset available, we can grow the trees without any synchronization hence faster.

**Comment by trivialfis** [comment]
I don't think we will implement this, you can split the multi-class problem into multiple binary classification problems in Python.
